core/src/services/icloud/session.rs

```diff
@@ -0,0 +1,304 @@
+use crate::{Error, ErrorKind};
+
+use crate::raw::oio::WriteBuf;
+use crate::raw::{new_json_deserialize_error, AsyncBody, HttpClient, IncomingAsyncBody};
+use crate::services::icloud::error::parse_error;
+use crate::services::icloud::webservices::iCloudWebservicesResponse;
+use serde_json::json;
+
+//TODO It just support cn
+//Other country shoule be https://www.icloud.com
+//Need to improve?
+const GLOBAL_HEADERS: [(&str, &str); 2] = [
+    ("Origin", "https://www.icloud.com.cn"),
+    ("Referer", "https://www.icloud.com.cn/"),
+];
+
+static ACCOUNT_COUNTRY_HEADER: &str = "X-Apple-ID-Account-Country";
+static OAUTH_STATE_HEADER: &str = "X-Apple-OAuth-State";
+
+static SESSION_ID_HEADER: &str = "X-Apple-ID-Session-Id";
+
+static SCNT_HEADER: &str = "scnt";
+
+static SESSION_TOKEN_HEADER: &str = "X-Apple-Session-Token";
+static AUTH_ENDPOINT: &str = "https://idmsa.apple.com/appleauth/auth";
+static SETUP_ENDPOINT: &str = "https://setup.icloud.com/setup/ws/1";
+
+static APPLE_RESPONSE_HEADER: &str = "X-Apple-I-Rscd";
+
+const AUTH_HEADERS: [(&str, &str); 7] = [
+    (
+        "X-Apple-OAuth-Client-Id",
+        "d39ba9916b7251055b22c7f910e2ea796ee65e98b2ddecea8f5dde8d9d1a815d",
+    ),
+    ("X-Apple-OAuth-Client-Type", "firstPartyAuth"),
+    ("X-Apple-OAuth-Redirect-URI", "https://www.icloud.com"),
+    ("X-Apple-OAuth-Require-Grant-Code", "true"),
+    ("X-Apple-OAuth-Response-Mode", "web_message"),
+    ("X-Apple-OAuth-Response-Type", "code"),
+    (
+        "X-Apple-Widget-Key",
+        "d39ba9916b7251055b22c7f910e2ea796ee65e98b2ddecea8f5dde8d9d1a815d",
+    ),
+];
+
+pub struct ServiceInfo {
+    pub(crate) url: String,
+}
+
+pub struct SessionData {
+    oauth_state: String,
+    session_id: Option<String>,
+    session_token: Option<String>,
+
+    scnt: Option<String>,
+    account_country: Option<String>,
+
+    cookies: BTreeMap<String, String>,
+    webservices: HashMap<String, ServiceInfo>,
+}
+
+impl SessionData {
+    pub fn new() -> SessionData {
+        Self {
+            oauth_state: format!("auth-{}", uuid::Uuid::new_v4().to_string()).to_string(),
+            session_id: None,
+            session_token: None,
+            scnt: None,
+            account_country: None,
+            cookies: BTreeMap::new(),
+            webservices: HashMap::new(),
+        }
+    }
+}
+
+pub struct Session {
+    pub(crate) data: SessionData,
+    pub(crate) client: HttpClient,
+    pub(crate) apple_id: String,
+    pub(crate) password: String,
+
+    pub trust_token: Option<String>,
+    pub ds_web_auth_token: Option<String>,
+}
+
+impl Debug for Session {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        let mut de = f.debug_struct("iCloud Session");
+        de.field("apple_id", &self.apple_id);
+        de.field("password", &self.password);
+        de.field("trust_token", &self.trust_token);
+        de.field("ds_web_auth_token", &self.ds_web_auth_token);
+        de.finish()
+    }
+}
+
+impl Session {
```

Comment:
`Session` is more like `Signer` in other services.

Reply:
I actually prefer `Session`.
```
pub struct IcloudCore {
    pub root: String,
    pub client: HttpClient,
    pub signer: Arc<IcloudSigner>,
}
```
or
```
pub struct IcloudCore {
    pub root: String,
    pub client: HttpClient,
    pub session: Arc<IcloudSession>,
}
```

Reply:
Please use Signer to make it easier to understand for other maintainers.